File: LeetCode/Medium/0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] answer = new int[n];
        
        // Step 1: Forward pass to compute left (prefix) products.
        // Initialize the first element to 1 as there are no elements to its left.
        answer[0] = 1;
        for (int i = 1; i < n; i++) {
            // Each position stores the product of the previous position's 
            // prefix product and the previous element in the input array.
            answer[i] = answer[i - 1] * nums[i - 1];
        }
        
        // Step 2: Backward pass to compute right (suffix) products on the fly.
        // rightProduct serves as an O(1) space accumulator.
        int rightProduct = 1;
        for (int i = n - 1; i >= 0; i--) {
            // Combine the left product (stored in answer) with the right product.
            answer[i] *= rightProduct;
            // Accumulate the current element into the right product for the next step.
            rightProduct *= nums[i];
        }
        
        return answer;
    }
}

File: LeetCode/Medium/0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
#include <vector>

class Solution {
public:
    std::vector<int> productExceptSelf(const std::vector<int>& nums) {
        int n = nums.size();
        // Initialize the answer vector with 1s to handle the boundaries implicitly.
        std::vector<int> answer(n, 1);
        
        // Step 1: Forward pass to compute left (prefix) products.
        // We start at index 1 and use the previously stored product.
        for (int i = 1; i < n; ++i) {
            answer[i] = answer[i - 1] * nums[i - 1];
        }
        
        // Step 2: Backward pass to compute right (suffix) products on the fly.
        // right_product is the scalar variable enabling the O(1) space constraint.
        int right_product = 1;
        for (int i = n - 1; i >= 0; --i) {
            // The final result for index i is left_product * right_product.
            answer[i] *= right_product;
            // Update the suffix accumulator.
            right_product *= nums[i];
        }
        
        return answer;
    }
};